File: evaluation/S0/java/Hard/1925-find-the-shortest-superstring.java
// Source: https://leetcode.com/problems/find-the-shortest-superstring/   |   Difficulty: Hard
//
// Problem Description:
// Given an array of strings words, return the smallest string that contains each string in words as a substring. If there are multiple valid strings of the smallest length, return any of them.
//
// You may assume that no string in words is a substring of another string in words.
//
// Example:
// Input: words = ["alex","loves","leetcode"]
// Output: "alexlovesleetcode"
// Explanation: All permutations of "alex","loves","leetcode" would also be accepted.
//
// Constraints:
// 1 <= words.length <= 12
// 	1 <= words[i].length <= 20
// 	words[i] consists of lowercase English letters.
// 	All the strings of words are unique.
//

class Solution {
    public String shortestSuperstring(String[] words) {
        int n = words.length;
        int[][] overlap = new int[n][n];

        // Precompute the overlap between each pair of words
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    int maxOverlap = Math.min(words[i].length(), words[j].length());
                    for (int k = maxOverlap; k >= 0; k--) {
                        if (words[i].endsWith(words[j].substring(0, k))) {
                            overlap[i][j] = k;
                            break;
                        }
                    }
                }
            }
        }

        // DP to find the shortest superstring
        int[][] dp = new int[1 << n][n];
        int[][] parent = new int[1 << n][n];

        for (int mask = 1; mask < (1 << n); mask++) {
            for (int bit = 0; bit < n; bit++) {
                if ((mask & (1 << bit)) == 0) continue;
                int prevMask = mask ^ (1 << bit);
                if (prevMask == 0) continue;

                for (int i = 0; i < n; i++) {
                    if ((prevMask & (1 << i)) == 0) continue;
                    int val = dp[prevMask][i] + words[bit].length() - overlap[i][bit];
                    if (val > dp[mask][bit]) {
                        dp[mask][bit] = val;
                        parent[mask][bit] = i;
                    }
                }
            }
        }

        // Reconstruct the shortest superstring
        int maxLen = -1;
        int last = -1;
        int finalMask = (1 << n) - 1;

        for (int i = 0; i < n; i++) {
            if (dp[finalMask][i] > maxLen) {
                maxLen = dp[finalMask][i];
                last = i;
            }
        }

        StringBuilder sb = new StringBuilder();
        boolean[] seen = new boolean[n];
        int mask = finalMask;

        while (mask > 0) {
            int prev = parent[mask][last];
            if (prev == 0 && (mask ^ (1 << last)) == 0) {
                sb.insert(0, words[last]);
            } else {
                sb.insert(0, words[last].substring(overlap[prev][last]));
            }
            seen[last] = true;
            mask ^= (1 << last);
            last = prev;
        }
        return sb.toString();
    }
}

File: evaluation/S0/python/Hard/1925-find-the-shortest-superstring.py
# Source: https://leetcode.com/problems/find-the-shortest-superstring/   |   Difficulty: Hard
#
# Problem Description:
# Given an array of strings words, return the smallest string that contains each string in words as a substring. If there are multiple valid strings of the smallest length, return any of them.
#
# You may assume that no string in words is a substring of another string in words.
#
# Example:
# Input: words = ["alex","loves","leetcode"]
# Output: "alexlovesleetcode"
# Explanation: All permutations of "alex","loves","leetcode" would also be accepted.
#
# Constraints:
# 1 <= words.length <= 12
# 	1 <= words[i].length <= 20
# 	words[i] consists of lowercase English letters.
# 	All the strings of words are unique.
#

class Solution(object):
    def shortestSuperstring(self, words):
        """
        :type words: List[str]
        :rtype: str
        """
        n = len(words)
        overlap = [[0] * n for _ in range(n)]
        
        for i in range(n):
            for j in range(n):
                if i != j:
                    max_ov = min(len(words[i]), len(words[j]))
                    for k in range(max_ov, 0, -1):
                        if words[i][-k:] == words[j][:k]:
                            overlap[i][j] = k
                            break
        
        dp = [[float('inf')] * n for _ in range(1 << n)]
        parent = [[-1] * n for _ in range(1 << n)]
        
        for i in range(n):
            dp[1 << i][i] = len(words[i])
        
        for mask in range(1 << n):
            for u in range(n):
                if mask & (1 << u):
                    for v in range(n):
                        if mask & (1 << v) == 0:
                            new_mask = mask | (1 << v)
                            cost = dp[mask][u] + len(words[v]) - overlap[u][v]
                            if cost < dp[new_mask][v]:
                                dp[new_mask][v] = cost
                                parent[new_mask][v] = u
        
        min_len = float('inf')
        last = -1
        final_mask = (1 << n) - 1
        
        for i in range(n):
            if dp[final_mask][i] < min_len:
                min_len = dp[final_mask][i]
                last = i
        
        path = []
        mask = final_mask
        while last != -1:
            path.append(last)
            temp = parent[mask][last]
            mask ^= (1 << last)
            last = temp
        
        path.reverse()
        
        result = words[path[0]]
        for i in range(1, len(path)):
            ov = overlap[path[i-1]][path[i]]
            result += words[path[i]][ov:]
        
        return result